prompt2model/prompt_parser/instr_parser.py

```diff
@@ -85,39 +86,29 @@ def parse_from_prompt(self, prompt: str) -> None:
         parsing_prompt_for_chatgpt = construct_prompt_for_instruction_parsing(prompt)
 
         chat_api = APIAgent()
+        last_error = None
         while True:
             self.api_call_counter += 1
-            response = chat_api.generate_one_completion(
-                parsing_prompt_for_chatgpt,
-                temperature=0,
-                presence_penalty=0,
-                frequency_penalty=0,
-            )
-
-            if isinstance(response, Exception):
-                # Generation failed due to an API related error and requires retry.
-
-                if self.max_api_calls and self.api_call_counter >= self.max_api_calls:
-                    # In case we reach maximum number of API calls, we raise an error.
-                    logger.error("Maximum number of API calls reached.")
-                    raise ValueError(
-                        "Maximum number of API calls reached."
-                    ) from response
-
-                continue  # no need to proceed with extracting
-                # response if API call failed.
-
-            extraction = self.extract_response(response)
-
-            if extraction is not None:
-                # extraction is successful
-
-                self._instruction, self._examples = extraction
-                return None
-
-            if self.max_api_calls and self.api_call_counter == self.max_api_calls:
-                # In case we reach maximum number of API calls without a
-                # successful extraction, we return None.
-
-                logger.warning("Maximum number of API calls reached for PromptParser.")
-                return None
+            try:
+                response: openai.ChatCompletion | Exception = (
+                    chat_api.generate_one_completion(
+                        parsing_prompt_for_chatgpt,
+                        temperature=0,
+                        presence_penalty=0,
+                        frequency_penalty=0,
+                    )
+                )
+                extraction = self.extract_response(response)
+                if extraction is not None:
+                    self._instruction, self._examples = extraction
+                    return
+            except API_ERRORS as e:
```

Comment:
Awesome, the changes simplify it further!
I wonder if catching `API_ERRORS` and calling `handle_api_error` in the parser or the generator might create issues later? Different providers (hf) might throw different errors from the API ones and may need different handling? That's why I was thinking error handling could be provider specific and part of the provider, and the retry logic can be part of the parser/generator classes.

Reply:
Totally agree. I usually don't like catching errors until the highest possible level of control flow, but we can design this in a future PR.